File: imperfect-anagrams.js
#!/usr/bin/env node --experimental-worker
const readline = require('readline');
const fs = require('fs');
const ProgressBar = require('progress');
const os = require('os');
const { Worker } = require('worker_threads');


/*
 * This script calculates which word contains the most other words within it, allowing for some letters to be skipped
 * i.e. then can spell the, he, hen & ten (notice that ten skips one letter)
 */

const words = [];

const trie = {
	children: {},
	words: []
};

let trieNodesCount = 0;

class WordDetails {
	constructor(word) {
		this.word = word;
		const sortedLetters = word.split('').sort();

		this.letterCounts = {};
		this.sortedSet = [sortedLetters[0]];

		for (let i = 0; i < sortedLetters.length; i++) {
			const curr = sortedLetters[i];

			if (this.letterCounts[curr]) {
				this.letterCounts[curr]++;
			} else {
				this.letterCounts[curr] = 1;
			}

			if (this.sortedSet[this.sortedSet.length - 1] !== curr) {
				this.sortedSet.push(curr);
			}
		}
	}
}

function run() {
	console.time('Worker Spawning');

	let maxWord;
	let maxWordsList = [];
	let workersReceivedFrom = 0;

	const numCpus = os.cpus().length;
	const chunkSize = Math.floor(words.length / numCpus);
	const reportFrequency =  Math.floor(chunkSize / 100);

	const bar = new ProgressBar('[:bar] :rate/wps :percent', {total: words.length});

	for (let i = 0; i < numCpus; i++) {
		const start = i * chunkSize;
		const end = i < numCpus - 1 ? (i + 1) * chunkSize : words.length;

		const worker = new Worker('./anagram-worker.js', {workerData: {trie, words, start, end}});
		worker.on('message', (data) => {
			if (data.tick >= 0) {
				bar.tick(data.tick, {});
				return;
			}

			if (data.anagrams) {
				workersReceivedFrom++;
				if (data.anagrams.length > maxWordsList.length) {
					maxWord = data.maxWord;
					maxWordsList = data.anagrams;
				}

				if (workersReceivedFrom === numCpus) {
					console.log(`The word "${maxWord}" has ${maxWordsList.length} imperfect anagrams`);
					if (maxWordsList.length > 100) {
						console.log('Words have been written to anagrams.txt');
						fs.writeFileSync('anagrams.txt', maxWordsList.join(', '));
					} else {
						console.log(maxWordsList.join(','));
					}
				}

				return;
			}

			console.error(`Unhandled message from worker`, data);
		});
		worker.on('error', error => console.error(`Error from worker ${i}`, error));
		worker.on('exit', code => {
			if (code !== 0) {
				console.error(`Worker ${i} exited with code ${code}`);
			}
		});

		let onlineCount = 0;
		worker.on('online', () => {
			onlineCount++;

			if (onlineCount === numCpus) {
				console.timeEnd('Worker Spawning');
			}
		});
	}
}

function addWord(word) {
	const wordDetails = new WordDetails(word);
	words.push(wordDetails);

	let head = trie;
	for (let j = 0; j < wordDetails.sortedSet.length; j++) {
		const letter = wordDetails.sortedSet[j];

		if (!head.children[letter]) {
			trieNodesCount++;
			head.children[letter] = {
				children: {},
				words: []
			};
		}

		head = head.children[letter];
	}

	if (Object.keys(head.children).length > 0) {
		wordDetails.canSkip = true;
	}

	head.words.push(wordDetails);
}

const lineReader = readline.createInterface({
	// input: fs.createReadStream('2000_random_words_alpha.txt')
	input: fs.createReadStream('first_2000_words.txt')
	// input: fs.createReadStream('words_alpha.txt')
});

console.time('Time to add');
lineReader.on('line', (word) => {
	addWord(word);
});

lineReader.on('close', function () {
	console.timeEnd('Time to add');
	console.log(`${(process.memoryUsage().heapUsed / 1024 / 1024).toFixed(1)} mb. ${trieNodesCount.toLocaleString()} nodes in trie`);
	run();
});

File: imperfect-anagrams.go
package main

import (
	"bufio"
	"fmt"
	"io"
	"log"
	"math"
	"os"
	"sort"
	"strings"
	"time"
)

type WordDetails struct {
	word         string
	sortedSet    []string
	letterCounts map[string]uint8
	canSkip      bool
}

type WordDetailsSlice []WordDetails

type Node struct {
	children map[string]*Node
	words    []*WordDetails
}

type Trie struct {
	root *Node
	size int
}

var words []WordDetails
var trie = Trie{
	root: &Node{children: make(map[string]*Node)},
	size: 0,
}

func newWordDetails(word string) WordDetails {
	var details = WordDetails{
		word,
		[]string{},
		make(map[string]uint8),
		false,
	}

	sortedLetters := strings.Split(word, "")
	sort.Strings(sortedLetters)

	details.sortedSet = append(details.sortedSet, sortedLetters[0])

	for _, char := range sortedLetters {
		details.letterCounts[char] += 1

		if details.sortedSet[len(details.sortedSet)-1] != char {
			details.sortedSet = append(details.sortedSet, char)
		}
	}

	return details
}

func timeTrack(start time.Time, name string) {
	elapsed := time.Since(start)
	fmt.Printf("%s took %s", name, elapsed)
}

func isWordAnagram(mainWord *WordDetails, wordToCheck *WordDetails) bool {
	if len(wordToCheck.word) > len(mainWord.word) {
		return false
	}

	for _, char := range wordToCheck.sortedSet {
		if wordToCheck.letterCounts[char] > mainWord.letterCounts[char] {
			return false
		}
	}

	return true
}

func searchSet(letters []string, start int, head *Node, currentWord *WordDetails, wordList *WordDetailsSlice) {
	if len(head.words) > 0 {
		for i := 0; i < len(head.words); i++ {
			if isWordAnagram(currentWord, head.words[i]) {
				*wordList = append(*wordList, *head.words[i])
			}
		}
	}

	for i := start; i < len(letters); i++ {
		if _, ok := head.children[letters[i]]; ok {
			searchSet(letters, i+1, head.children[letters[i]], currentWord, wordList)
		}
	}
}

type wordAnagramsPair struct {
	word     string
	anagrams []WordDetails
}

func findMostAnagramsForSlice(ticks chan<- int, anagrams chan<- wordAnagramsPair, wordSlice []WordDetails) {
	var maxWord string
	var maxAnagrams []WordDetails

	pingFrequency := len(wordSlice) / 100

	for i, currentWord := range wordSlice {
		if currentWord.canSkip {
			continue
		}

		if i % pingFrequency == 0 {
			ticks <- pingFrequency
		}

		var anagrams WordDetailsSlice
		searchSet(currentWord.sortedSet, 0, trie.root, &currentWord, &anagrams)

		// todo refactor searchSet to be iterative instead of recursive
		//word, anagrams := findAnagrams(currentWord)

		if len(anagrams) > len(maxAnagrams) {
			maxWord = currentWord.word
			maxAnagrams = anagrams
		}
	}

	if len(words) % pingFrequency != 0 {
		ticks <- len(words) % pingFrequency
	}
	anagrams <- wordAnagramsPair{maxWord, maxAnagrams}
}

func printProgress(ticks <-chan int) {
	start := time.Now()
	totalTicks := 0
	total := float64(len(words))

	for {
		amount := <-ticks
		if amount == -1 {
			return
		}

		totalTicks += amount
		curr := float64(totalTicks)

		ratio := math.Min(math.Max(curr/total, 0), 1)
		percent := int32(math.Floor(ratio * 100))
		elapsed := float64(time.Since(start).Seconds())
		eta := elapsed * (total/curr - 1)
		rate := curr / elapsed

		fmt.Printf("\r%d%% %d/wps %f seconds    ", percent, int32(rate), eta)
		//default:
		//	time.Sleep(100 * time.Millisecond)
		//}
	}
}

func findMostAnagrams() (string, []WordDetails) {
	const numCpus = 4
	chunkSize := len(words) / numCpus

	ticks := make(chan int, 50)
	defer close(ticks)
	anagrams := make(chan wordAnagramsPair)
	defer close(anagrams)

	go printProgress(ticks)

	for i := 0; i < numCpus; i++ {
		start := i * chunkSize
		end := (i + 1) * chunkSize
		if i == numCpus-1 {
			end = len(words)
		}

		wordSlice := words[start:end]

		go findMostAnagramsForSlice(ticks, anagrams, wordSlice)
	}

	fmt.Println()
	receivedCount := 0
	var maxAnagramsPair wordAnagramsPair

	for {
		select {
		case pair := <-anagrams:
			if len(pair.anagrams) > len(maxAnagramsPair.anagrams) {
				maxAnagramsPair = pair
			}
			receivedCount++

			if receivedCount == numCpus {
				ticks <- -1
				return maxAnagramsPair.word, maxAnagramsPair.anagrams
			}

		default:
			time.Sleep(100 * time.Millisecond)
		}
	}
}

func buildTrie(filename string) {
	defer timeTrack(time.Now(), "Build trie and word details")

	nodeCount := 0

	fileHandle, err := os.Open(filename)
	if err != nil {
		log.Fatal(err)
	}
	defer Close(fileHandle)

	scanner := bufio.NewScanner(fileHandle)
	for scanner.Scan() {
		var details WordDetails
		details = newWordDetails(scanner.Text())

		var head *Node
		head = trie.root

		for _, letter := range details.sortedSet {
			if _, ok := head.children[letter]; !ok {
				nodeCount++

				head.children[letter] = &Node{
					make(map[string]*Node),
					[]*WordDetails{},
				}
			}

			head = head.children[letter]
		}

		if len(head.children) > 0 {
			details.canSkip = true
		}

		words = append(words, details)
		head.words = append(head.words, &details)
	}

	if err := scanner.Err(); err != nil {
		log.Fatal(err)
	}

	fmt.Println("Trie nodes: ", nodeCount)
}

func main() {
	buildTrie("./words_alpha.txt")
	//buildTrie("./first_2000_words.txt")
	word, anagrams := findMostAnagrams()

	fmt.Printf("\rLongest word: %s with %d anagrams    ", word, len(anagrams))
}

func Close(c io.Closer) {
	err := c.Close()
	if err != nil {
		log.Fatal(err)
	}
}
